### PR TITLE
FSType is now optional in Volume objects and 'ext4' used when it's when omitted.

### PR DESCRIPTION
**Pull request by jsafrane**
This is follow up of https://github.com/kubernetes/kubernetes/pull/17650#discussion_r45755077.

Most volume plugins use `SafeFormatAndMount`, which uses ext4 by default.

`FlexVolume` plugin has `FSType` attribute 'omitempty', so reflect it in the
description of the type.

Also, update AWS EBS plugin to have the same style as the rest of plugins - see https://github.com/kubernetes/kubernetes/pull/17650#discussion_r45618690

@swagiaal, would you mind a short review?

### CODE REVIEW

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 1]
```diff
@@ -110,7 +110,7 @@ func (plugin *awsElasticBlockStorePlugin) newBuilderInternal(spec *volume.Spec,
 		fsType:      fsType,
 		partition:   partition,
 		readOnly:    readOnly,
-		diskMounter: &mount.SafeFormatAndMount{plugin.host.GetMounter(), exec.New()}}, nil
+		diskMounter: &mount.SafeFormatAndMount{mounter, exec.New()}}, nil
```
Comment (swagiaal) [inline pkg/volume/aws_ebs/aws_ebs.go]: Could you separate this change into its own PR it is not related to this PR.

**Comment by swagiaal** [comment]
@jsafrane thanks for following up...
The validation changes look good to me.
The api doc changes look good to but you will have to do them in v1/types.go as well.
The aws ebs change looks good but it should be in a separate PR

**Comment by jsafrane** [comment]
> you will have to do them in v1/types.go as well.

Will do shortly.

> The aws ebs change looks good but it should be in a separate PR

Separate commit is not enough?

**Comment by swagiaal** [comment]
> Separate commit is not enough?

I think it should be a separate PR. Separate commits are for related sequential changes building up to one big modification.

**Comment by jsafrane** [comment]
Ok, I removed AWS changes from the PR. I guess we'll sort it out when we rewrite plugins with the new Attach interface.

**Comment by swagiaal** [comment]
Hmm.. the aws change seems to still be there
Doc changes look good

**Comment by jsafrane** [comment]
Weird, I would swear I removed the patch. Now it's really removed.

**Comment by swagiaal** [comment]
@jsafrane thanks... LGTM

File: pkg/api/types.go [hunk 2]
```diff
@@ -502,9 +502,9 @@ type FCVolumeSource struct {
 type FlexVolumeSource struct {
 	// Driver is the name of the driver to use for this volume.
 	Driver string `json:"driver"`
-	// Required: Filesystem type to mount.
+	// Filesystem type to mount.
 	// Must be a filesystem type supported by the host operating system.
-	// Ex. "ext4", "xfs", "ntfs"
+	// Ex. "ext4", "xfs", "ntfs". The default one depends on FlexVolume script.
```
Comment (markturansky) [inline pkg/api/types.go]: "The default filesystem depends on ..."

File: pkg/api/v1/types.go [hunk 3]
```diff
@@ -663,9 +663,9 @@ type GCEPersistentDiskVolumeSource struct {
 type FlexVolumeSource struct {
 	// Driver is the name of the driver to use for this volume.
 	Driver string `json:"driver"`
-	// Required: Filesystem type to mount.
+	// Filesystem type to mount.
 	// Must be a filesystem type supported by the host operating system.
-	// Ex. "ext4", "xfs", "ntfs"
+	// Ex. "ext4", "xfs", "ntfs". The default one depends on FlexVolume script.
```
Comment (markturansky) [inline pkg/api/v1/types.go]: "The default filesystem depends on ..."

**Comment by markturansky** [comment]
LGTM after the wording changes above.

@bgrant0607 please review for API changes?

**Comment by bgrant0607** [comment]
This isn't strictly backward compatible. What's the motivation?

**Comment by markturansky** [comment]
None of these API fields are strictly required as all plugins involved in this PR use SafeFormatAndMount which defaults to "ext4".

Jan is looking to make the API reflect the reality of the code.

**Comment by swagiaal** [comment]
In addition all the fields say "omitempty"

File: pkg/api/v1/types.go (hunk 3 above)
Comment (jsafrane) [inline pkg/api/v1/types.go]: Updated, thanks.

**Comment by jsafrane** [comment]
@bgrant0607, as Mark wrote, motivation is to unify all the volumes. All of them already use the same code on the lower levels to format and mount a block device, this patch propagates this change to higher levels, i.e. type definition (add `omitempty` where applicable), validation and documentation.

**Comment by jsafrane** [comment]
Rebased and updated description according to @markturansky suggestions.

File: pkg/api/types.go (hunk 2 above)
Comment (jsafrane) [inline pkg/api/types.go]: Updated, thanks.

**Comment by jsafrane** [comment]
Rebased

**Comment by bgrant0607** [comment]
cc @thockin

**Comment by bgrant0607** [comment]
@markturansky The field is not set by default AFAICT:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/defaults.go

**Comment by markturansky** [comment]
@bgrant0607 the default value is found in the mount utils. Perhaps the wrong place, but you'll find it here:  https://github.com/kubernetes/kubernetes/blob/master/pkg/util/mount/mount_linux.go#L279

**Comment by bgrant0607** [comment]
@markturansky You're right. Not the right place. That won't be reflected in the API. When we say a field has a "default value", that value is represented explicitly when the user does a GET on the resource.

**Comment by bgrant0607** [comment]
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#defaulting

**Comment by bgrant0607** [comment]
If the omitempty changes reflect the code, it's better to make them consistent, but we'll need a release note to warn users about the incompatible API spec change, which may affect auto-generated clients.

The comments can't say "Defaults to...", because that's not true. It's more "Implicitly inferred to be...".

Please file an issue to fix this to comply with standard defaulting behavior in a future API version.

**Comment by jsafrane** [comment]
Rebased and comments updated to `Implicitly inferred to be "ext4"`. How do I update release notes?

**Comment by swagiaal** [comment]
> Please file an issue to fix this to comply with standard defaulting behavior in a future API version.

@bgrant0607 I'll take care of this: https://github.com/kubernetes/kubernetes/issues/20332

**Comment by bgrant0607** [comment]
Re. release note: I labeled this PR appropriately. Please draft a release note and put it in the PR description.

**Comment by bgrant0607** [comment]
Please add "if unspecified"

**Comment by jsafrane** [comment]
> Please add "if unspecified"

Added, now the comment says: `Implicitly inferred to be "ext4" if unspecified.`

**Comment by jsafrane** [comment]
rebased

**Comment by bgrant0607** [comment]
LGTM
